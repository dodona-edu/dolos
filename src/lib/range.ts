export class Range {
  private lowerBound: number;
  private upperBound: number;

  constructor(from: number, to: number) {
    if (from > to) {
      throw RangeError("from must be lower then to");
    }
    this.lowerBound = from;
    this.upperBound = to;
  }

  public get from(): number {
    return this.lowerBound;
  }

  public get to(): number {
    return this.upperBound;
  }

  public equals(other: Range): boolean {
    return this.to === other.to && this.from === other.from;
  }

  /**
   * Tests wether or not the given number can extend the range.
   * @param value The number you want to test.
   * @param gapSize The maximum gap size you want to allow. For example if the gap size is 0 then 2 cannot extend [4,5].
   * If you take a gap size of 1 however then the range can be extended and would become [2,5].
   */
  public canExtendWithNumber(value: number, gapSize: number = 0): boolean {
    return this.from - 1 - gapSize <= value && value <= this.to + 1 + gapSize;
  }

  public includes(value: number): boolean {
    return this.from <= value && value <= this.to;
  }

  /**
   * Extends the range with the given number.
   * @param value
   * @returns A reference to this.
   */
  public extendWithNumber(value: number): this {
    this.lowerBound = Math.min(value, this.from);
    this.upperBound = Math.max(value, this.to);
    return this;
  }

  /**
   * Extends the range with the given range in place. The given range will not be changed.
   * @param range The range you want this range to be extended by.
   */
  public extendWithRange(range: Range): this {
    this.lowerBound = Math.min(range.from, this.from);
    this.upperBound = Math.max(range.to, this.to);
    return this;
  }

  /**
   * Tests wether or not the given range can extend the current range.
   * @param range The range you want to test if can be used to extend the current range with.
   * @param gapSize The maximum gap size you want to allow. For example if the gap size is 0 then [1, 3] and [5,6]
   * will not be joined. If you take a gap size of 1 however then the ranges can be combined into [1,6].
   * @returns A boolean if the range can be used to extend the current range.
   */
  public canExtendWithRange(range: Range, gapSize: number = 0): boolean {
    const test =
      this.canExtendWithNumber(range.from, gapSize) ||
      this.canExtendWithNumber(range.to, gapSize) ||
      range.canExtendWithNumber(this.to, gapSize) ||
      range.canExtendWithNumber(this.from, gapSize);
    return test;
  }

  /**
   * Tests if the two ranges overlap.//TODO
   * @param other The other range
   */
  public overlappingLinesAmount(other: Range): number {
    if (!this.canExtendWithRange(other, 0)) {
      return 0;
    }

    if (Math.sign(this.from - other.from) !== Math.sign(this.to - other.to)) {
      return Math.min(this.getLineCount(), other.getLineCount());
    } else {
      return Math.min(this.to - other.from, other.to - this.from) + 1;
    }
  }

  public toString(): string {
    return `[${this.from + 1}, ${this.to + 1}]`;
  }

  /**
   * @returns The amount of lines in this range.
   */
  public getLineCount(): number {
    return this.to - this.from + 1;
  }
}
