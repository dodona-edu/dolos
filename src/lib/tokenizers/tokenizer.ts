import { default as fsWithCallbacks } from "fs";
const fs = fsWithCallbacks.promises;

export interface Token<Location> {
  token: string;
  location: Location;
}

export abstract class Tokenizer<Location> {
  /**
   * Returns a stringified version the given file.
   *
   * @param fileName The name of the file to parse
   */
  public async tokenizeFile(fileName: string): Promise<string> {
    const fileContent = await fs.readFile(fileName);
    return this.tokenize(fileContent);
  }

  /**
   * Returns a stringified version of the buffer
   *
   * @param text The buffer to stringify
   */
  public abstract tokenize(text: Buffer): string;

  /**
   * Runs the stringifier on a file with the given name.  Returns a tuple containing the
   * stringified version and an array containing a mapping from each token to the
   * corresponding token in the original buffer.
   *
   * @param fileName The name of the file to stringify
   */
  public async tokenizeFileWithMapping(fileName: string): Promise<[string, Location[]]> {
    const fileContent = await fs.readFile(fileName);
    return this.tokenizeWithMapping(fileContent);
  }

  /**
   * Runs the stringifier on a given buffer. Returns a tuple containing the stringified version
   * and an array containing a mapping from each token to the corresponding token in the
   * original buffer.
   *
   * @param text The text buffer to stringify
   */
  public async tokenizeWithMapping(text: Buffer): Promise<[string, Location[]]> {
    let resultString = "";
    const positionMapping: Location[] = [];
    for (const { token, location } of this.generateTokens(text)) {
      resultString += token;
      positionMapping.push(...new Array(token.length).fill(location));
    }
    return [resultString, positionMapping];
  }

  /**
   * Runs the stringifier on a file with the given name. Returns an async iterator returning
   * tuples containing the stringified version of the token and the corresponding position.
   *
   * @param fileName The name of the file to stringify
   */
  public async *generateTokensFromFile(fileName: string): AsyncIterableIterator<Token<Location>> {
    const fileContent = await fs.readFile(fileName);
    yield* this.generateTokens(fileContent);
  }

  /**
   * Runs the tokenizer on a given Buffer. Returns an async iterator returning tuples
   * containing the stringified version of the token and the corresponding position.
   *
   * @param text The text string to parse
   */
  public abstract generateTokens(text: Buffer): IterableIterator<Token<Location>>;

  /**
   * Returns a new token-object. Just a shorthand for {token: ..., location: ...}.
   *
   * @param token the text of the token
   * @param location the location of the token
   */
  protected newToken(token: string, location: Location): Token<Location> {
    return { token, location };
  }
}
