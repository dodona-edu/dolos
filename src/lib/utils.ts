/*
 * This module contains shared helper functions.
 */

import File from "./files/file";
import { Range } from "./range";

export type Colour = "red" | "green" | "reset";
export type RangesTuple = [Range, Range];
export type Clustered<T> = T[][];
export type Match = [File, File, RangesTuple[]];

export interface ObjectMap<T> {
  [key: string]: T;
}

/**
 * Colours your text with the given colour. Only works for terminal output.
 * @param colour The colour you want your text to be.
 * @param text The text you want to colour.
 */
export function colour(c: Colour, text: string): string {
  return `${escapeSeq(c)}${text}${escapeSeq("reset")}`;
}

/**
 * Helper function to convert a Colour 'enum' into its ANSI escape sequence.
 */
function escapeSeq(c: Colour) {
  switch (c) {
    case "red": return "\u001b[31m";
    case "green": return "\u001b[32m";
    case "reset": return "\u001b[0m";
  }
}

/**
 * Counts the total amount of lines that correspond with the first and second file.
 * @param rangesTupleArray The rangesTupleArray you want the lines count of.
 * @returns A tuple that contais the line for the first and second file respectively.
 */
export function countLinesInRanges(rangesTupleArray: RangesTuple[]): [number, number] {
  return rangesTupleArray
    .map(([range1, range2]) => [range1.getLineCount(), range2.getLineCount()] as [number, number])
    .reduce(([acc1, acc2], [next1, next2]) => [acc1 + next1, acc2 + next2]);
}

export function optionsToString(optionsArray: Array<[string, string | number]>): string {
  return (
    optionsArray
    .map(([flag, optionValue]) => {
      if (typeof optionValue === "string" && optionValue.includes(" ")) {
        return [flag, `'${optionValue}'`];
      }
      return [flag, optionValue];
    })
    .map(([flag, optionValue]) => `${flag} ${optionValue}`)
    .join(" ") + "\n"
  );
}

export function sortRangesTuples([r11, r12]: RangesTuple, [r21, r22]: RangesTuple): number {
  const diff = r11.from - r21.from;
  if (diff === 0) {
    return r12.from - r22.from;
  } else {
    return diff;
  }
}

export function scoreForFiles(
    matches: RangesTuple[],
    matchedFile: File,
    matchingFile: File,
  ): [number, number] {

    const [matchedLinesInMatchedFile, matchedLinesInMatchingFile]: [
      number,
      number,
    ] = countLinesInRanges(matches);

    const linesInMatchedFile: number = matchedFile.lineCount.okOr(0);
    const linesInMatchingFile: number = matchingFile.lineCount.okOr(0);
    const scoreMatchedFile: number = Math.round(
      (matchedLinesInMatchedFile / linesInMatchedFile) * 100,
    );
    const scoreMatchingFile: number = Math.round(
      (matchedLinesInMatchingFile / linesInMatchingFile) * 100,
    );

    return [scoreMatchedFile, scoreMatchingFile];
  }
