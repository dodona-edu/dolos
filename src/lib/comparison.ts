import { HashFilter } from "./hashFilter";
import { NoFilter } from "./noFilter";
import { Tokenizer } from "./tokenizer";
import { WinnowFilter } from "./winnowFilter";

export type Matches<Location> = Map<string, Array<[Location, Location]>>;

/**
 * @param filterFragmentsByPercentage Defines if the fragment should be filtered by percentage or by an absolute value.
 * @param maxFragment The maximum fragment. How this will be used depends on the value of [[filterFragmentByPercentage]]. //TODO better wording needed
 * If it is used as a percentage then the number should be between 0 and 1. If you want to use it as an absolute value
 * then a number between 0 and the amount of files given would be the most useful.
 */
export interface ComparisonFilterOptions {
  maxFragment: number;
  filterFragmentByPercentage: boolean;
}

/**
 * @param hashFilter An optional HashFilter to filter the hashes returned by
 * the rolling hash function.
 * @param noFilter A NoFilter used to generate hashes for blacklisted files.
 */
export interface ComparisonOptions<Location> {
  tokenizer: Tokenizer<Location>;
  hashFilter?: HashFilter;
  noFilter?: NoFilter;
  filterOptions?: ComparisonFilterOptions;
}

export class Comparison<Location> {
  private readonly defaultK: number = 50;
  private readonly defaultW: number = 40;
  // Maps a hash to an array of typles that are composed of the file with the same hash and the location in that file
  // where that hash is located.
  private readonly index: Map<number, Array<[string, Location]>> = new Map();
  private readonly filteredHashSet: Set<number> = new Set();
  private readonly tokenizer: Tokenizer<Location>;
  private readonly hashFilter: HashFilter;
  private readonly noFilter: NoFilter;
  private readonly fragmentFilterOptions: ComparisonFilterOptions | undefined;
  private fileCount: number = 0;

  /**
   * Creates a Comparison object with a given Tokenizer and optional HashFilter.
   * If no HashFilter is given, a new WinnowFilter with k-mer length 50 and windows
   * size 40 will be used.
   *
   * After creation, first add files to the index which can then be queried.
   *
   * @param tokenizer A tokenizer for the correct programming language
   * @param fragmentOptions The options used to filter based on the fragment count. For a more detailed explanation see
   * [[ComparisonOptions]]. If this options is not used then no filtering will occur.
   */
  constructor({ tokenizer, hashFilter, noFilter, filterOptions }: ComparisonOptions<Location>) {
    this.tokenizer = tokenizer;
    this.hashFilter = hashFilter || new WinnowFilter(this.defaultK, this.defaultW);
    this.noFilter = noFilter || new NoFilter(this.defaultK);
    this.fragmentFilterOptions = filterOptions;
  }

  /**
   * Adds a list of files to the comparison index.
   *
   * @param files A list of filenames
   */
  public addFiles(files: string[]): Promise<void[]> {
    // This promise will reject if one of the underlying promises reject
    // which is not what we want. If one file is missing, the others should
    // still be added. In the future, Promise.all can be replaced by
    // Promise.allSettled, but for now the error handling makes sure they
    // are all resolved correctly.
    return Promise.all(files.map(file => this.addFile(file)));
  }

  /**
   * Add a file to the comparison index.
   *
   * @param file The file name of the file to add
   */
  public async addFile(file: string): Promise<void> {
    this.fileCount += 1;
    try {
      const [ast, mapping] = await this.tokenizer.tokenizeFileWithMapping(file);
      for await (const { hash, location } of this.hashFilter.hashesFromString(ast)) {
        // hash and the corresponding line number //
        const match: [string, Location] = [file, mapping[location]];
        const matches = this.index.get(hash);
        if (matches) {
          matches.push(match);
        } else {
          this.index.set(hash, [match]);
        }
      }
    } catch (error) {
      console.error(`There was a problem parsing ${file}. ${error}`);
      return; // this makes sure the promise resolves instead of rejects
    }
  }

  /**
   * Compare a list of files to the index. A map will be returned containing for each
   * file in the input a list of matching files. This list of matching files is also
   * represented as a map, mapping the filename to the position of the match in both files.
   *
   * @param files A list of filenames to query
   * @param hashFilter An optional HashFilter. By default the HashFilter of the Comparison
   * object will be used.
   */
  public async compareFiles(
    files: string[],
    hashFilter = this.hashFilter,
  ): Promise<Map<string, Matches<Location>>> {
    const matchingFiles: Map<string, Matches<Location>> = new Map();
    for (const file of files) {
      matchingFiles.set(file, await this.compareFile(file, hashFilter));
    }
    return matchingFiles;
  }

  /**
   * Compare a file to the index. A map will be returned containing the filename
   * of the matching file, along with a list of matching position between the two files.
   *
   * @param file The file to query
   * @param hashFilter An optional HashFilter. By default the HashFilter of the Comparison
   * object will be used.
   */
  public async compareFile(file: string, hashFilter = this.hashFilter): Promise<Matches<Location>> {
    // mapping file names to line number pairs (other file, this file)
    const matchingFiles: Matches<Location> = new Map();
    const [ast, mapping] = await this.tokenizer.tokenizeFileWithMapping(file);
    for await (const { hash, location } of hashFilter.hashesFromString(ast)) {
      if (this.filteredHashSet.has(hash)) {
        continue;
      }
      const matches = this.index.get(hash);
      if (!matches || !this.filterByFragmentCount(matches.length)) {
        continue;
      }

      for (const [fileName, lineNumber] of matches) {
        // add the match if the match is not with the file and if the file comes first when alphabetically sorted.
        // This is done to avoid the duplicates in the following case: when file A matches with file B, file B will
        // also match with file A.
        if (fileName === file) {
          continue;
        }
        const match: [Location, Location] = [lineNumber, mapping[location]];
        const lines = matchingFiles.get(fileName);
        if (lines) {
          lines.push(match);
        } else {
          matchingFiles.set(fileName, [match]);
        }
      }
    }
    return matchingFiles;
  }

  /**
   * Add a file to the filter index.
   * @param file The file name of the file you want to add.
   */
  public async addFileToFilterList(file: string): Promise<void> {
    try {
      const [ast] = await this.tokenizer.tokenizeFileWithMapping(file);
      for await (const { hash } of this.noFilter.hashesFromString(ast)) {
        this.filteredHashSet.add(hash);
      }
    } catch (error) {
      console.error(`There was a problem parsing ${file}. ${error}`);
      return; // this makes sure the promise resolves instead of rejects
    }
  }

  /**
   *
   * @param fragmentCount The amount of times a certain code fragment has appeared
   * @returns true if is value is acceptable under the options given in the contructor or when the options are not
   * defined.
   */
  private filterByFragmentCount(fragmentCount: number): boolean {
    if (!this.fragmentFilterOptions) {
      return true;
    } else if (this.fragmentFilterOptions.filterFragmentByPercentage) {
      return fragmentCount / this.fileCount <= this.fragmentFilterOptions.maxFragment;
    } else {
      return fragmentCount <= this.fragmentFilterOptions.maxFragment;
    }
  }
}
