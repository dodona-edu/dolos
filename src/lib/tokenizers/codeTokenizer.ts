import { default as Parser, SyntaxNode } from "tree-sitter";
import { Token, Tokenizer } from "./tokenizer";

export class CodeTokenizer extends Tokenizer<number> {
  public static supportedLanguages = ["c-sharp", "haskell", "java",
    "javascript", "python"];

  /**
   * Returns true if the grammar of the given language is supported.
   *
   * @param language The name of the language to check
   */
  public static isSupportedLanguage(language: string): boolean {
    return this.supportedLanguages.includes(language);
  }

  /**
   * Registers an additional language to Dolos. For this to work, the supporting
   * module of the name `tree-sitter-someLanguage` must first be installed
   * manually through yarn or npm.
   *
   * The function will throw an error when the supported module is not found.
   *
   * @param language The name of the language to register
   */
  public static registerLanguage(language: string): void {
    try {
      require("tree-sitter-" + language);
    } catch (error) {
      throw new Error(
        `The module 'tree-sitter-${language}' could not be found`
      );
    }
    this.supportedLanguages.push(language);
  }

  public readonly language: string;
  private readonly parser: Parser;

  /**
   * Creates a new tokenizer of the given language. Will throw an error when the
   * given language is not supported. See Tokenizer.supportedLanguages for a
   * list of all supported languages.
   *
   * @param language The language to use for this tokenizer.
   */
  constructor(language: string) {
    super();
    if (!CodeTokenizer.isSupportedLanguage(language)) {
      throw new Error(`Language '${language}' is not supported`);
    }

    this.language = language;
    this.parser = new Parser();
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const languageModule = require("tree-sitter-" + language);
    this.parser.setLanguage(languageModule);
  }

  /**
   * Runs the parser on a given string. Returns a stringified version of the
   * abstract syntax tree.
   *
   * @param text The text string to parse
   */
  public tokenize(text: string): string {
    const tree = this.parser.parse(text);
    return tree.rootNode.toString();
  }

  /**
   * Runs the parser on a given string. Returns an async iterator returning
   * tuples containing the stringified version of the token and the
   * corresponding position.
   *
   * @param text The text string to parse
   */
  public *generateTokens(text: string): IterableIterator<Token<number>> {
    const tree = this.parser.parse(text);
    yield* this.tokenizeNode(tree.rootNode);
  }

  private *tokenizeNode(node: SyntaxNode): IterableIterator<Token<number>> {
    const location = node.startPosition.row;

    yield this.newToken("(", location);
    // "(node.type child1 child2 ...)"
    for (const c of node.type) {
      yield this.newToken(c, location);
    }

    for (const child of node.namedChildren) {
      yield this.newToken(" ", location);
      yield* this.tokenizeNode(child);
    }
    yield this.newToken(")", location);
  }
}
