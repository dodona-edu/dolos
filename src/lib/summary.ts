import fs from "fs";
import { Matches } from "./comparison";
import { Range } from "./range";
type RangesTuple = [Range, Range];
export class Summary {
  private readonly results: Map<string, Matches<Range>>;
  private readonly minimumLines: number;
  private readonly gapSize: number;

  /**
   * @param matches A many-to-many comparison of a set of files. This map contains an entry for each of the
   * input files with the key being its file name and the value a list of matches. These matches are grouped
   * per matching file. The compareFiles function of the Comparison class can generate such mapping.
   * @param minimumLines The minimum amount of lines required by a rangesTuple. If the rangesTuple has less lines then 
   * it will be filtered out. When the rangesTuple has two ranges with a different amount of lines, then the maximum
   * between to two is used.
   * @param gapSize The gap size allowed during the joining of two ranges. For example if the gap size is 0 then [1,3]
   * and [5,7] wont be joined, and if the gap size is one these will be joined into [1,7].
   */
  constructor(
    matchesPerFile: Map<string, Matches<number>>,
    minimumLines: number = 2,
    gapSize: number = 1,
  ) {
    this.minimumLines = minimumLines;
    this.gapSize = gapSize;
    this.results = this.transformMatches(matchesPerFile);
    this.results = this.sortResults();
  }

  public toString(): string {
    if (this.results.size === 0) {
       return "There were no matches";
    }

    return Array.from(this.results.entries()).map((resultEntry) => {
      let [sourceFileName, subMap] = resultEntry;
      let output = "";
      output += `source: ${sourceFileName}\n\n`;
      const linesInSourceFile = this.countLinesInFile(sourceFileName);

      output += Array.from(subMap.entries()).map((subMapEntry) => {
        let matchedFilenameOutput = "";
        let [matchedFileName, rangesTupleArray] = subMapEntry;
        const score: number = rangesTupleArray
          .map(rangesTuple => rangesTuple[0].getLineCount())
          .reduce((accumulator, nextValue) => accumulator + nextValue);

        const scoreMatchedFile = (score / this.countLinesInFile(matchedFileName)) * 100;
        const scoreSourceFile = (score / linesInSourceFile) * 100;

        matchedFilenameOutput += `\tmatched file: ${matchedFileName}, score matched file: ${Math.round(
          scoreMatchedFile,
        )}%, score source file: ${Math.round(scoreSourceFile)}%\n\n`;

        matchedFilenameOutput += "\tranges: ";
        matchedFilenameOutput += rangesTupleArray.map(rangesTuple => `[${rangesTuple[0]}, ${rangesTuple[1]}]`);
        matchedFilenameOutput += "\n\n";
        return matchedFilenameOutput;
      }).join('');
      return output
    }).join('');
  }

  /**
   * Checks pairwise if the first element of each RangesTuple can be extended with the second.
   * @param rangesTuple1 The tuple where the ranges will be tested if it can be extended from the ranges from the 
   * second tuple.
   * @param rangesTuple2 The tuple where the ranges will be used to extend with.
   */
  public canExtentRangesTupleWithRangesTuple(
    rangesTuple1: RangesTuple,
    rangesTuple2: RangesTuple,
  ): boolean {
    return (
      rangesTuple1[0].canExtendWithRange(rangesTuple2[0]) &&
      rangesTuple1[1].canExtendWithRange(rangesTuple2[1])
    );
  }

  /**
   * Attempts the extend the first element of each tuple with each other and tries the same for the second element.
   * Assumes that it is possible to perform this operation, an error is thrown if this is not the case.
   * @param rangesTuple1 The rangesTuple where the ranges will be extended.
   * @param rangesTuple2 The rangesTuple where the ranges will be used to extend.
   */
  public extendRangesTupleWithRangesTuple(rangesTuple1: RangesTuple, rangesTuple2: RangesTuple) {
    if (
      this.canExtentRangesTupleWithRangesTuple(rangesTuple1, rangesTuple2)
    ) {
      throw new RangeError("a value in the rangeTuple could not be extended");
    }

    rangesTuple1[0].extendWithRange(rangesTuple2[0]);
    rangesTuple1[1].extendWithRange(rangesTuple2[1]);
  }

  private countLinesInFile(fileName: string): number {
    return fs.readFileSync(fileName, "utf8").split("\n").length;
  }

  /**
   * Calculates the score, currently just returns the number of lines in the range. A possible alternative is counting
   * the number of k-mers.
   * @param range The range you want to get the score of
   * @returns The score
   */
  private getScoreForRange(range: Range): number {
    return range.getLineCount();
  }

  /**
   * First sorts the array of rangesTuples, then the subMaps and finally the main maps according to their corresponding
   * score functions.
   */
  private sortResults(): Map<string, Matches<Range>> {
    // TODO index the score of the ranges, arrays and submaps to make this more efficient.
    this.results.forEach((matches, matchedFileName) => {
      matches.forEach((rangesTupleArray, _) => {
        // Sorts the arrays based on the score of the ranges.
        rangesTupleArray.sort(
          (rangesTuple1, rangesTuple2) =>
            this.getScoreForRange(rangesTuple2[0]) - this.getScoreForRangesTuple(rangesTuple1),
        );
      });
      // Sorts the submaps based on the score of the arrays, this is the sum of all the scores within the array.
      const tempMatches = new Map(
        [...matches.entries()].sort(
          (subMapEntry1, subMapEntry2) =>
            this.getScoreForArray(subMapEntry2[1]) - this.getScoreForArray(subMapEntry1[1]),
        ),
      );
      this.results.set(matchedFileName, tempMatches);
    });

    // sorts the maps based on the score of the submaps, which is the sum of the scores contained within the submaps.
    return new Map(
      [...this.results.entries()].sort(
        (subMap1, subMap2) =>
          this.getScoreForSubMap(subMap2[1]) - this.getScoreForSubMap(subMap1[1]),
      ),
    );
  }

  /**
   * Transforms all the tuples to rangesTuples.
   * @param matchesPerFile A many-to-many comparison of a set of files. This map contains an entry for each of the
   * input files with the key being its file name and the value a list of matches. These matches are grouped
   * per matching file. The compareFiles function of the Comparison class can generate such mapping.
   */
  private transformMatches(
    matchesPerFile: Map<string, Matches<number>>,
  ): Map<string, Matches<Range>> {
    const results: Map<string, Matches<Range>> = new Map();
    matchesPerFile.forEach((matches, matchedFileName) => {
      matches.forEach((matchLocations, matchingFile) => {
        let map: Matches<Range> | undefined = results.get(matchingFile);
        const rangesTupleArray: RangesTuple[] = this.matchesToRange(matchLocations);
        if (rangesTupleArray.length !== 0) {
          if (map === undefined) {
            map = new Map();
            results.set(matchedFileName, map);
          }
          map.set(matchingFile, rangesTupleArray);
        }
      });
    });
    return results;
  }

  private getScoreForArray(arr: RangesTuple[]): number {
    return arr
      .map(rangesTuple => this.getScoreForRangesTuple(rangesTuple))
      .reduce((acc, nextNumber) => acc + nextNumber);
  }

  private getScoreForSubMap(subMap: Matches<Range>): number {
    return [...subMap.values()]
      .map(rangesArray => this.getScoreForArray(rangesArray))
      .reduce((acc, nextNumber) => acc + nextNumber);
  }

  private getScoreForRangesTuple(rangesTuple: RangesTuple): number {
    return this.getScoreForRange(rangesTuple[0]) + this.getScoreForRange(rangesTuple[1]);
  }

  /**
   * Converts a list of matching lines to a list of ranges.
   * @param matches The list of matching lines.
   * @returns A list of tuples that contains two ranges, where the frist and second range correspond to the line
   * numbers of each file.
   */
  private matchesToRange(matches: Array<[number, number]>): RangesTuple[] {
    const ranges: RangesTuple[] = new Array();

    matches.forEach(next => {
      const rangeTupleIndex: number = ranges.findIndex(rangeTuple => {
        return (
          rangeTuple[0].canExtendWithNumber(next[0]) && rangeTuple[1].canExtendWithNumber(next[1])
        );
      });

      if (rangeTupleIndex === -1) {
        ranges.push([new Range(next[0], next[0], this.gapSize), new Range(next[1], next[1], this.gapSize)]);
      } else {
        ranges[rangeTupleIndex][0].extendWithNumber(next[0]);
        ranges[rangeTupleIndex][1].extendWithNumber(next[1]);
      }
    });

    // If two rangesTuples overlap with each other then extend the second with the first and remove the
    // first from the array.
    for (let i = ranges.length - 1; i >= 0; i--) {
      for (let j = i; j >= 0; j--) {
        if (i !== j) {
          // console.log(ranges[i], ranges[j], newRangesTuple);
          if (this.canExtentRangesTupleWithRangesTuple(ranges[i], ranges[j])) {
            this.extendRangesTupleWithRangesTuple(ranges[j], ranges[i]);
            ranges.splice(i, 1);
          }
          break;
        }
      }
    }

    // Remove all ranges that only contain less the minimum required lines.
    return ranges.filter(
      rangesTuple =>
        Math.max(rangesTuple[0].getLineCount(), rangesTuple[1].getLineCount()) >=
        this.minimumLines,
    );
  }
}
