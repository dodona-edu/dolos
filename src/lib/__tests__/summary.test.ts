import { Comparison, Matches } from "../comparison";
import { Range } from "../range";
import { FilterOptions, Summary } from "../summary";
import { CodeTokenizer } from "../tokenizers/codeTokenizer";
import { RangesTuple } from "../utils";

/**
 * adapted from https://stackoverflow.com/questions/6274339/how-can-i-shuffle-an-array
 * @param a An array containing the items.
 */

function shuffle<T>(a: T[]) {
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

let seed = 1;
function random(): number {
  const x = Math.sin(seed++) * 10000;
  return x - Math.floor(x);
}

test("simple match to ranges", () => {
  const summary = new Summary(new Map());
  const array: Array<[number, number]> = [[1, 5], [2, 6], [3, 7], [4, 8]];
  shuffle(array);

  const rangesTupleArray: RangesTuple[] = summary.matchesToRange(array);
  expect(rangesTupleArray).toContainEqual([new Range(1, 4), new Range(5, 8)]);
  expect(rangesTupleArray.length).toBe(1);
});

test("simple match to ranges where second element stays constant", () => {
  const summary = new Summary(new Map());
  const array: Array<[number, number]> = [[1, 5], [2, 5], [3, 5], [4, 5]];
  shuffle(array);

  const rangesTupleArray: RangesTuple[] = summary.matchesToRange(array);
  expect(rangesTupleArray).toContainEqual([new Range(1, 4), new Range(5, 5)]);
  expect(rangesTupleArray.length).toBe(1);
});

test("simple match to ranges where first element stays constant", () => {
  const summary = new Summary(new Map());
  const array: Array<[number, number]> = [[1, 5], [1, 6], [1, 7], [1, 8]];
  shuffle(array);

  const rangesTupleArray: RangesTuple[] = summary.matchesToRange(array);
  expect(rangesTupleArray).toContainEqual([new Range(1, 1), new Range(5, 8)]);
  expect(rangesTupleArray.length).toBe(1);
});

test("simple match to ranges where first element stays constant and a gap exists", () => {
  const summary1 = new Summary(new Map(), 1);
  const summary2 = new Summary(new Map(), 0);
  const array: Array<[number, number]> = [[1, 5], [1, 6], [1, 7], [1, 9]];
  shuffle(array);

  const summary2Array = summary2.matchesToRange(array);
  const summary1Array = summary1.matchesToRange(array);
  expect(summary1Array).toContainEqual([new Range(1, 1), new Range(5, 9)]);
  expect(summary1Array.length).toBe(1);
  expect(summary2Array).toContainEqual([new Range(1, 1), new Range(5, 7)]);
  expect(summary2Array).toContainEqual([new Range(1, 1), new Range(9, 9)]);
  expect(summary2Array.length).toBe(2);
});

test("test extending related functions rangesTuples", () => {
  const summary1 = new Summary(new Map(), 0);
  const summary2 = new Summary(new Map(), 1);
  const rangesTuple1: RangesTuple = [new Range(1, 5), new Range(1, 5)];
  const rangesTuple2: RangesTuple = [new Range(6, 10), new Range(6, 20)];
  const rangesTuple3: RangesTuple = [new Range(7, 10), new Range(7, 30)];
  const rangesTuple4: RangesTuple = [new Range(1, 5), new Range(1, 5)];
  const rangesTuple5: RangesTuple = [new Range(12, 200), new Range(21, 5000)];

  expect(summary1.canExtendRangesTupleWithRangesTuple(rangesTuple1, rangesTuple2)).toBe(true);
  expect(summary1.canExtendRangesTupleWithRangesTuple(rangesTuple1, rangesTuple3)).toBe(false);
  expect(summary2.canExtendRangesTupleWithRangesTuple(rangesTuple1, rangesTuple3)).toBe(true);

  summary1.extendRangesTupleWithRangesTuple(rangesTuple1, rangesTuple2);
  expect(rangesTuple1).toEqual([new Range(1, 10), new Range(1, 20)]);
  summary2.extendRangesTupleWithRangesTuple(rangesTuple4, rangesTuple3);
  expect(rangesTuple4).toEqual([new Range(1, 10), new Range(1, 30)]);
  expect(() => summary1.extendRangesTupleWithRangesTuple(rangesTuple1, rangesTuple5)).toThrowError(
    RangeError,
  );
});

test("concatenate ranges", () => {
  const summary = new Summary(new Map());
  const rangesTupleArray: RangesTuple[] = [
    [new Range(0, 10), new Range(100, 110)],
    [new Range(10, 30), new Range(110, 120)],
    [new Range(5, 25), new Range(105, 115)],
    [new Range(555, 2001), new Range(0, 33)],
  ];
  const shuffledArray1 = rangesTupleArray.slice();
  const shuffledArray2 = rangesTupleArray.slice();

  shuffle(shuffledArray1);
  shuffle(shuffledArray2); // The second array will have a different order then the first.

  const concatenatedRanges1 = summary.concatenateRanges(shuffledArray1);
  const concatenatedRanges2 = summary.concatenateRanges(shuffledArray2);

  expect(concatenatedRanges1).toContainEqual([new Range(0, 30), new Range(100, 120)]);
  expect(concatenatedRanges1).toContainEqual(rangesTupleArray[3]);
  expect(concatenatedRanges1).not.toContainEqual([new Range(5, 25), new Range(105, 115)]);

  expect(concatenatedRanges2).toContainEqual([new Range(0, 30), new Range(100, 120)]);
  expect(concatenatedRanges2).toContainEqual(rangesTupleArray[3]);
  expect(concatenatedRanges2).not.toContainEqual([new Range(5, 25), new Range(105, 115)]);
});

test("integration test", async () => {
  const locations: string[] = [
    "samples/js/sample.js",
    "samples/js/copied_function.js",
    "samples/js/another_copied_function.js",
  ];
  const tokenizer = new CodeTokenizer("javascript");
  const comparison = new Comparison(tokenizer, {
    filterHashByPercentage: true,
    maxHash: 0.8,
  });
  comparison.addFiles(locations);

  const matchesPerFile: Map<string, Matches<number>> = await comparison.compareFiles(locations);

  const filterOptions: FilterOptions = {
    minimumFragmentLength: 4,
  };
  const summary = new Summary(matchesPerFile, 2, filterOptions, 2);

  expect(summary.toString()).toMatchSnapshot();
  expect(summary.toJSON()).toMatchSnapshot();
});
