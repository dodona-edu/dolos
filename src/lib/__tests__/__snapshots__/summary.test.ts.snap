// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration test 1`] = `
"Cluster 1
	samples/js/sample.js(7%) + samples/js/copied_function.js(36%) => 	
		[[4,7],[15,18]]
	samples/js/copied_function.js(36%) + samples/js/sample.js(7%) => 
		[[15,18],[4,7]]

"
`;

exports[`integration test 2`] = `"{\\"results\\":[[[\\"samples/js/sample.js\\",\\"samples/js/copied_function.js\\",[[[4,7],[15,18]]]],[\\"samples/js/copied_function.js\\",\\"samples/js/sample.js\\",[[[15,18],[4,7]]]]]]}"`;
